Port `/h` and `/s` to RzSearch.

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Ports the `/s` (calculate blocks entropies and select those above a threshold) and `/h` to RzSearch and RzShell command `/ce`.

- Add `/ch` command for hash search and an `/ce` alias for `entropy_fract` search.
- Adds the block size as command to `/ch`. So the command no loner depends on `core->blocksize`.
- Extends `RzSearchHit` with `details`. They can be set arbitrary by the search and printed as well. For the entropy search this is the fractured entropy value.


**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Added

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

wargio commented on librz/include/rz_search.h:
```diff
@@ -213,13 +231,31 @@ typedef enum {
 	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_SAFECURVES,
 	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_X509,
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_END_CRYPTO,
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_ALL_KEY_BASED, ///< All key based algos from above.
+
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_BEGIN_HASH, ///< Hash related algorithms
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_ENTROPY, ///< Block entropies.
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_HASH, ///< Block hashes.
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_END_HASH,
+
```
TBH, i would have a different collection for hash and entropy.

Rot127 replied:
I thought about it as well. But decided entropy alone is is too simple to get its own collection. Especially, since there are not additions. Also it fits neatly into the same design of the hash search. The enum is weird though, waited what you would say.
Maybe we can just extract the hash + entropy into a separated collection, so they share one?

But I would like to keep the passing of data to the `find()` workers. Because we might add crypto algos which require it.

Rot127 replied:
Never mind, I remove the find() data from the crypto search. There is no need to have dead code for "maybe usage".

wargio commented on librz/include/rz_search.h:
```diff
@@ -146,9 +147,48 @@ typedef struct rz_search_interval_t RzSearchInterval;
+	RZ_SEARCH_COLLECTION_HASH_SHA1,
+	RZ_SEARCH_COLLECTION_HASH_SM3,
+	RZ_SEARCH_COLLECTION_HASH_XOR,
+	RZ_SEARCH_COLLECTION_HASH_CRC,
```
There are many CRC algos. I would just push the name to RzHash instead of enums.

wargio commented on librz/include/rz_search.h:
```diff
@@ -213,14 +255,22 @@ typedef enum {
 	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_ECC,
 	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_SAFECURVES,
 	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_X509,
+	RZ_SEARCH_COLLECTION_CRYPTOGRAPHIC_ALL, ///< All algos from above.
```
I would use a #define for this setting it to enum_size

wargio commented on librz/include/rz_search.h:
```diff
@@ -213,14 +255,22 @@ typedef enum {
 } RzSearchCollectionCryptographicType;
 
-RZ_API RZ_OWN RzSearchCollection *rz_search_collection_cryptographic(bool add_all_methods);
+RZ_API RZ_OWN RzSearchCollection *rz_search_collection_cryptographic();
```
Revert this?

Rot127 replied:
I'd like to keep it this way. We have `rz_search_collection_cryptographic_add` for adding methods. Adding search method in the initialization function would delude the consistency of the design.
Meaning: After initialization of the any collection, it is expected to be empty. Search methods have to be added in a separated step.

But I extract this change into a separated PR in a few minutes.

Rot127 replied:
https://github.com/rizinorg/rizin/pull/5002

wargio commented on librz/search/hash_search.c:
```diff
@@ -0,0 +1,389 @@
+	rz_return_val_if_fail(expected_digits, NULL);
+	ut8 *out = RZ_NEWS0(ut8, digit_size);
+	switch (algo) {
+	case RZ_SEARCH_COLLECTION_HASH_ENUM_SIZE:
```
This should be default

wargio reviewed (changes requested):
For the entropy I would forbid it in the hash command and have 3 new commands to which checks if entropy is less than/equal/greater than

Rot127 committed:
Mark RzHash as const to ensure thread safety.

This is important because it is required to setup a hasher.
RzHash is normally initialized with all hash-plugins and owned by RzCore.
So, if several threads need to hash data, RzHash becomes shared data.
And immutability has to be ensured.

Rot127 committed:
Make error message of failed boundary init more explicit.

Rot127 committed:
Move initialization of all crypto search methods to the add() function.

Rot127 committed:
Allow adding arbitrary details to RzSearchHit.

Rot127 committed:
Port /h and /s to RzShell and RzSearch.

- Adds the new command /ch for hash searches and the alias /ce for /s.
- Adds search hit details to the entropy hits.

Rot127 commented:
> For the entropy I would forbid it in the hash command and have 3 new commands to which checks if entropy is less than/equal/greater than

Equal wouldn't work too well due to rounding errors. The less than case is nice. But can I open an issue about this please? I have too many other tasks to take care of currently.

wargio commented:
Yes. If you want we can merge it as is and I can just fix this with a pr.
Potentially this can be just a search between x and y values.

wargio commented:
Ah this probably should be implemented for entropy and fractional entropy

Rot127 commented:
> Yes. If you want we can merge it as is and I can just fix this with a pr.
> Potentially this can be just a search between x and y values.
> Ah this probably should be implemented for entropy and fractional entropy

Sounds good to me.

wargio commented:
Please rebase
